src/scales/scale.time.js

```diff
@@ -779,27 +779,36 @@ module.exports = Scale.extend({
 	 */
 	getLabelCapacity: function(exampleTime) {
 		var me = this;
-		var timeOpts = me.options.time;
+		var options = me.options;
+		var timeOpts = options.time;
 		var displayFormats = timeOpts.displayFormats;
-		var margins = me.margins;
 
 		// pick the longest format (milliseconds) for guestimation
 		var format = displayFormats[timeOpts.unit] || displayFormats.millisecond;
 		var exampleLabel = me.tickFormatFunction(exampleTime, 0, ticksFromTimestamps(me, [exampleTime], me._majorUnit), format);
 		var size = me._getLabelSize(exampleLabel);
+		var labelSize, scaleSize, chartSize;
 
-		// Using margins instead of padding because padding is not calculated
-		// at this point (buildTicks). Margins are provided from previous calculation
-		// in layout steps 5/6
-		var capacity = Math.floor(me.isHorizontal()
-			? (me.width - margins.left - margins.right) / size.w
-			: (me.height - margins.top - margins.bottom) / size.h);
+		if (me.isHorizontal()) {
+			labelSize = size.w;
+			scaleSize = me.width;
+			chartSize = me.chart.width;
+		} else {
+			labelSize = size.h;
+			scaleSize = me.height;
+			chartSize = me.chart.height;
+		}
 
-		if (me.options.offset) {
-			capacity--;
+		// If offset is true, the scale size is reduced by the half label size on each edge.
+		// If false, it is reduced if the chart doesn't have enough margin for the half label
+		// size on one edge and by the half size on the other edge. See #6300 for details.
+		if (options.offset) {
+			scaleSize -= labelSize;
+		} else {
+			scaleSize = Math.min(scaleSize, chartSize - labelSize / 2) - labelSize / 2;
```

Comment:
If there are no y-axes, full label size should be reduced (this is accounted for here) and if there are y-axes both sides (with room for label extending that direction), then no label size should be reduced, right?.
The problem is, margins are already reduced from width if there are axes there. Didn't actually check if its dependent on position also (so if left margin is reduced, but right is not, even if there is a y-axis with `position: 'right'`).

So `scaleSize` could end up being `labelSize / 2` too small here. [example](https://jsfiddle.net/xojz5r1q/)

Another issue is, if there are no y-axes, but x-axes top and bottom and labels are rotated. Then `labelSize` is not enough: [example](https://jsfiddle.net/nt38cjsL/)

I really think scale should not be using `chartSize` for anything, because it does (and should) not know enough. 

All that said, this works a lot better than master / #6115

Reply:
> If there are no y-axes, full label size should be reduced (this is accounted for here) and if there are y-axes both sides (with room for label extending that direction), then no label size should be reduced, right?.

That's correct.

> The problem is, margins are already reduced from width if there are axes there.

This is not true. `getLabelCapacity` is first called through `update` in Step 4 then called again in Step 5&6 during layout. The margins are not passed on the first call, so they don't affect the width. They are reduced from the width in `fit` which is called after `buildTick`, so they don't affect in the second call, either.

> Didn't actually check if its dependent on position also (so if left margin is reduced, but right is not, even if there is a y-axis with position: 'right').

Right, there is no direct way to know on which side a y-axis is, so the this PR code is simply assuming it is on either side.

> So scaleSize could end up being labelSize / 2 too small here. [example](https://jsfiddle.net/xojz5r1q/)

I think this example gives a not bad result if you change the canvas width to 270.
![Screen Shot 2019-05-27 at 5 34 32 PM](https://user-images.githubusercontent.com/723188/58410915-c2f05400-80a5-11e9-9e28-6bf1e4eff895.png)

> Another issue is, if there are no y-axes, but x-axes top and bottom and labels are rotated. Then labelSize is not enough: [example](https://jsfiddle.net/nt38cjsL/)

This is actually a very difficult problem because a scale doesn't know label widths in other scales.

> I really think scale should not be using chartSize for anything, because it does (and should) not know enough.

I agree that ideally it should not. This is also a compromise solution.